pkg/apis/operator/v1alpha1/tower_types.go

```diff
@@ -0,0 +1,232 @@
+				s.Phase = TowerPhaseTerminating
+			} else {
+				s.Phase = TowerPhaseNotReady
+			}
+		case ConditionUnknown:
+			s.Phase = TowerPhaseUnknown
+		}
+		return
+	}
+
+	s.Phase = TowerPhaseUnknown
+}
+
+// TowerConditionType represents a TowerCondition value.
+type TowerConditionType string
+
+const (
+	// TowerReady represents a Tower condition is in ready state.
+	TowerReady TowerConditionType = "Ready"
+)
+
+// TowerCondition contains details for the current condition of this Tower.
+type TowerCondition struct {
+	// Type is the type of the Tower condition, currently ('Ready').
+	Type TowerConditionType `json:"type"`
+	// Status is the status of the condition, one of ('True', 'False', 'Unknown').
+	Status ConditionStatus `json:"status"`
+	// LastTransitionTime is the last time the condition transits from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	// Reason is the (brief) reason for the condition's last transition.
+	Reason string `json:"reason"`
+	// Message is the human readable message indicating details about last transition.
+	Message string `json:"message"`
+}
+
+// True returns whether .Status == "True"
+func (c TowerCondition) True() bool {
+	return c.Status == ConditionTrue
+}
+
+// GetCondition returns the Condition of the given condition type, if it exists.
+func (s *TowerStatus) GetCondition(t TowerConditionType) (condition TowerCondition, exists bool) {
+	for _, cond := range s.Conditions {
+		if cond.Type == t {
+			condition = cond
+			exists = true
+			return
+		}
+	}
+	return
+}
+
+// SetCondition replaces or adds the given condition.
+func (s *TowerStatus) SetCondition(condition TowerCondition) {
+	defer s.updatePhase()
+
+	if condition.LastTransitionTime.IsZero() {
+		condition.LastTransitionTime = metav1.Now()
+	}
+
+	for i := range s.Conditions {
+		if s.Conditions[i].Type == condition.Type {
+
+			// Only update the LastTransitionTime when the Status is changed.
+			if s.Conditions[i].Status != condition.Status {
+				s.Conditions[i].LastTransitionTime = condition.LastTransitionTime
+			}
+
+			s.Conditions[i].Status = condition.Status
+			s.Conditions[i].Reason = condition.Reason
+			s.Conditions[i].Message = condition.Message
+
+			return
+		}
+	}
+
+	s.Conditions = append(s.Conditions, condition)
+}
+
+// Tower manages the deployment of the KubeCarrier master controller manager.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.phase"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type Tower struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TowerSpec   `json:"spec,omitempty"`
+	Status TowerStatus `json:"status,omitempty"`
+}
+
+// IsReady returns if the Tower is ready.
+func (s *Tower) IsReady() bool {
+	if s.Generation != s.Status.ObservedGeneration {
+		return false
+	}
+
```

Comment:
Please add check if the tower is being deleted then it's unready. Maybe now would be a good time to autogenerate all this common functionality for our components.